### PR TITLE
BLUEBUTTON-1281: Jenkins Build, Update and Deploy

### PR DESCRIPTION
**Pull request by jzulim**
### Change Details

This PR attempts to improve our jenkins building, updating and deployment process.

#### Deployment
- TF - Adds jenkins_boundary policy to jenkins role
- TF - Via a EFS module, Provisions EFS and mount point, to replace EBS JENKINS_HOME. 
- TF - Updated to reflect the need for two environments, mgmt and mgmt-test. 
- TF - Removes extra Jenkins permissions / groups, not needed. Simplified. 
- TF - Moves to launch templates for ASG's. 
- Packer - Updates jenkins build and config packer templates to include additional var (env). 
- Ansible - Rekeys jenkins vault with new vault pass

#### Building the Jenkins AMI
- Secured and segmented ssh access for all engineers via ssh_users role
- Installs and configures the Cloudwatch agent. 
- Installs ops tools (aws, jq, terraform via tfenv, ansible, pip) via ops_tools role
- Adds paramterized-scheduler plugin
- Installs the swarm plugin
- Bumps instance size to c5.xlarge
- Configured lockable-resource plugin (lock envs) 
- Mounts EFS for JENKINS_HOME (/var/lib/jenkins)
- Updates github token (rotated)
- Removes proxy (healthapt leftover) 
- removes epel_repo role, not needed. 
- Installs nfs-utils for EFS mounting. 

#### Update Jenkins
- Simplified plugin update task 
- Adds EFS mount role. 

### Acceptance Validation

I've built, updates and deployed to both the mgmt-test and mgmt environments using this new code. Will want to peer review this PR, show output, answer questions. It's a lot to take in and I'm happy to do an interactive review at first pass and let reviewing engineers then take it away. 

### External References

- [BLUEBUTTON-1281](https://jira.cms.gov/browse/BLUEBUTTON-1281)

### Security Implications

- [ ] new software dependencies

- [ ] altered security controls

- [ ] new data stored or transmitted

- [ ] security checklist is completed for this change

- [ ] requires more information or team discussion to evaluate security implications

### CODE REVIEW

**Comment by jzulim** [comment]
Updated README to refer to the need to STOP jenkins before build and deploying. Also confirmed the swarm plugin is installed by running another update_jenkins packer build. 

`amazon-ebs: changed: [default] => (item=swarm)`

**Comment by jzulim** [comment]
`   amazon-ebs: PLAY RECAP *********************************************************************
    amazon-ebs: default                    : ok=84   changed=62   unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
    amazon-ebs:
    amazon-ebs: Thursday 05 March 2020  07:04:29 -0800 (0:00:07.509)       0:20:36.985 ********
    amazon-ebs: ===============================================================================
    amazon-ebs: ops_tools : Install latest version of Ansible via pip ----------------- 143.82s
    amazon-ebs: install_docker : Add Jenkins User to Docker Group --------------------- 135.05s
    amazon-ebs: install_jenkins : Install Plugins ------------------------------------- 111.08s
    amazon-ebs: base_patch : Apply security patches ----------------------------------- 106.93s
    amazon-ebs: install_docker : Install Docker ---------------------------------------- 64.24s
    amazon-ebs: rhel_prep : install the latest version of ntp -------------------------- 56.17s
    amazon-ebs: install_jenkins : Install - AWS CLI ------------------------------------ 26.43s
    amazon-ebs: ssh_users : Add user public key ---------------------------------------- 24.69s
    amazon-ebs: ssh_users : Add user --------------------------------------------------- 24.05s
    amazon-ebs: install_jenkins : Install Pre-requisites ------------------------------- 23.79s
    amazon-ebs: ssh_users : Add user to sudoers file ----------------------------------- 23.19s
    amazon-ebs: build_install_apache : Update Configuration Files ---------------------- 22.61s
    amazon-ebs: build_install_apache : SELinux - Allow apache to modify files in /var/www/builds -- 21.81s
    amazon-ebs: rhel_python_pip : Ensure botocore and boto3 modules are installed ------ 18.74s
    amazon-ebs: ops_tools : Install latest version of git ------------------------------ 18.57s
    amazon-ebs: install_jenkins : Wait for Jenkins HTTP OK ----------------------------- 16.36s
    amazon-ebs: install_jenkins : Jenkins Install -------------------------------------- 15.17s
    amazon-ebs: build_install_apache : Install Apache ---------------------------------- 14.68s
    amazon-ebs: install_jenkins : Configure Jenkins Launch Settings -------------------- 13.58s
    amazon-ebs: build_install_apache : SELinux - Allow Apache to Proxy ----------------- 12.10s
    amazon-ebs:
    amazon-ebs:
    amazon-ebs: Log written to: /Users/johnzulim/repo/cms/bfd/beneficiary-fhir-data/ops/ansible/playbooks-ccs/logs/ansible-production-2020-03-05T06:43:49-0800.log
==> amazon-ebs: Stopping the source instance...
    amazon-ebs: Stopping instance
==> amazon-ebs: Waiting for the instance to stop...
==> amazon-ebs: Creating AMI bfd-jenkins-20200305024201 from instance i-0c76e31a7e44c7c8f
    amazon-ebs: AMI: ami-0fd47765e06480bc8
==> amazon-ebs: Waiting for AMI to become ready...`

**Comment by jzulim** [comment]
```Plan: 1 to add, 1 to change, 1 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.stateless.module.jenkins.aws_autoscaling_group.main: Destroying... [id=bfd-mgmt-test-jenkins-6]
module.stateless.module.jenkins.aws_launch_template.main: Modifying... [id=lt-0645f0550c2db5287]
module.stateless.module.jenkins.aws_launch_template.main: Modifications complete after 5s [id=lt-0645f0550c2db5287]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 10s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 20s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 30s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 40s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 50s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 1m0s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 1m10s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 1m20s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 1m30s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still destroying... [id=bfd-mgmt-test-jenkins-6, 1m40s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Destruction complete after 1m47s
module.stateless.module.jenkins.aws_autoscaling_group.main: Creating...
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [10s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [20s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [30s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [40s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [50s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m0s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m10s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m20s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m30s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m40s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [1m50s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [2m0s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Still creating... [2m10s elapsed]
module.stateless.module.jenkins.aws_autoscaling_group.main: Creation complete after 2m10s [id=bfd-mgmt-test-jenkins-8]

Apply complete! Resources: 1 added, 1 changed, 1 destroyed.
Releasing state lock. This may take a few moments...```

File: ops/terraform/modules/resources/efs/main.tf [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  description = "EFS Security Group"
+  vpc_id      = data.aws_vpc.main.id
+
+  // NFS
+  ingress {
+    description     = "Inbound access for EFS"
+    from_port = 2049
+    to_port = 2049
+    protocol = "tcp"
+    cidr_blocks = ["10.252.40.0/21"]
```
[the remark this answers is not in the document]
Comment (jzulim) [reply ops/terraform/modules/resources/efs/main.tf]: `cidr_blocks = ["${data.aws_vpc.main.cidr_block}"]`

Maybe something like that.

File: ops/ansible/playbooks-ccs/roles/mount_efs/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+---
+##
+# Mounts Jenkins Data EBS Volume
+##
+- name: Ensure NFS is installed.
+  package: "name={{ nfs_package }} state=installed"
```
[the remark this answers is not in the document]
Comment (jzulim) [reply ops/ansible/playbooks-ccs/roles/mount_efs/tasks/main.yml]: I'm sorry, I'm not following. Can you show what you mean?

File: ops/terraform/modules/mgmt_stateful/main.tf [hunk 3]
```diff
@@ -26,6 +26,16 @@ data "aws_subnet_ids" "app_subnets" {
   }
 }
 
+# Subnet for EFS 
+
+data "aws_subnet_ids" "efs" {
+  vpc_id = data.aws_vpc.main.id
+
+  tags = {
+    Name = "bfd-mgmt-az3-app"
```
[the remark this answers is not in the document]
Comment (jzulim) [reply ops/terraform/modules/mgmt_stateful/main.tf]: I think I have a way. Hold please. :)

File: ops/ansible/playbooks-ccs/roles/mount_efs/tasks/main.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (njdister) [reply ops/ansible/playbooks-ccs/roles/mount_efs/tasks/main.yml]:
```
package:
  name={{ nfs_package }}
  state=installed
```

File: ops/terraform/modules/mgmt_stateful/main.tf (hunk 3 above)
[the remark this answers is not in the document]
Comment (jzulim) [reply ops/terraform/modules/mgmt_stateful/main.tf]: Turns out I dont think this is needed. In the EFS module I grab the vpc and subnet data via: 

```data "aws_subnet" "main" {
  vpc_id            = data.aws_vpc.main.id
  availability_zone = var.env_config.azs
  filter {
    name    = "tag:Layer"
    values  = [var.layer] 
  }
}```

**Pull request merged**